XenAdmin/TabPages/PvsPage.cs

```diff
@@ -0,0 +1,208 @@
+
+                if (dataGridViewFarms.SelectedRows.Count == 0 && dataGridViewFarms.Rows.Count > 0)
+                    dataGridViewFarms.Rows[0].Selected = true;
+            }
+            finally
+            {
+                dataGridViewFarms.ResumeLayout();
+            }
+        }
+
+        private void LoadVMs()
+        {
+            Program.AssertOnEventThread();
+
+            if (!Visible)
+                return;
+
+            try
+            {
+                dataGridViewVms.SuspendLayout();
+                
+                UnregisterVMHandlers();
+                dataGridViewVms.Rows.Clear();
+                
+                foreach (var pvsProxy in Connection.Cache.PVS_proxies.Where(p => p.VM != null))
+                    dataGridViewVms.Rows.Add(NewVmRow(pvsProxy));
+
+                if (dataGridViewVms.SelectedRows.Count == 0 && dataGridViewVms.Rows.Count > 0)
+                    dataGridViewVms.Rows[0].Selected = true;
+            }
+            finally
+            {
+                dataGridViewVms.ResumeLayout();
+            }
+        }
+
+        private DataGridViewRow NewPvsFarmRow(PVS_farm pvsFarm)
+        {
+            var farmCell = new DataGridViewTextBoxCell {Value = pvsFarm.name};
+            var configurationCell = new DataGridViewTextBoxCell
+            {
+                Value = pvsFarm.cache_storage.Count > 0 ? Messages.PVS_CACHE_MEMORY_AND_DISK: Messages.PVS_CACHE_MEMORY_ONLY
+            };
+            var cacheSrsCell = new DataGridViewTextBoxCell
+            {
+                Value = string.Join(",  ", Connection.ResolveAll(pvsFarm.cache_storage))
+            };
+
+            var newRow = new DataGridViewRow { Tag = pvsFarm };
+            newRow.Cells.AddRange(farmCell, configurationCell, cacheSrsCell);
+
+            return newRow;
+        }
+
+        private DataGridViewRow NewVmRow(PVS_proxy pvsProxy)
+        {
+            var vm = pvsProxy.VM;
+            System.Diagnostics.Trace.Assert(vm != null);
+            var vmCell = new DataGridViewTextBoxCell {Value = pvsProxy.VM.Name};
+            var cachedCell = new DataGridViewTextBoxCell
+            {
+                Value = pvsProxy.currently_attached ? Messages.YES : Messages.NO
+            };
+            var srCell = new DataGridViewTextBoxCell {Value = Connection.Resolve(pvsProxy.cache_SR)};
+            var prepopulationCell = new DataGridViewTextBoxCell
+            {
+                Value = pvsProxy.prepopulate ? Messages.YES : Messages.NO
+            };
+
+            var newRow = new DataGridViewRow { Tag = vm };
+            newRow.Cells.AddRange(vmCell, cachedCell, srCell, prepopulationCell);
+            vm.PropertyChanged += VmPropertyChanged;
+
+            return newRow;
+        }
+
+        private void UnregisterVMHandlers()
+        {
+            foreach (DataGridViewRow row in dataGridViewVms.Rows)
+            {
+                var vm = row.Tag as VM;
+                if (vm == null) 

```

Comment:
Is this null check needed (or the one on line 203)? Can the VM ever be null here? On line 122 we only add a row when the VM for a PVS Proxy is not null, so is it possible for the VM here to become null after being added as a row?